src/astral/page-script.js

```diff
@@ -0,0 +1,200 @@
+    const originalRequest = request.request;
+    if (originalRequest.rollAttack && !originalRequest.rollDamage) {
+        rollDamages = `beyond20-rendered-roll-button-${Math.random()}`;
+        rolls.push(["Roll Damages", `[\`Click\`](#${rollDamages})`]);
+    }
+    if (originalRequest.type === "initiative" && settings["initiative-tracker"]) {
+        const initiative = request.attack_rolls.find((roll) => !roll.discarded);
+        if (initiative)
+           rolls.push(["Initiative", `i!(d0cr=1fr=1 + ${initiative.total})`]);
+    }
+    let message = template(rolls);
+    await postChatMessage({characterName: request.character, message, ...getDecoration(request.request.type), title, whisper: request.whisper == WhisperType.YES });
+    if (rollDamages) {
+        await hookRollDamages(rollDamages, request);
+    }
+}
+
+async function postChatMessage({characterName, message, color, icon, title, whisper}) {
+    try {
+        const user = getUser().uid;
+        const room = getRoom();
+        const token = await getAccessToken();
+        const character = await getCharacter(characterName);
+
+        const recipients = whisper ? Object.fromEntries(getDungeonMasters().map(key => [key, true])) : undefined;
+        if (message.length > 2048) {
+            message = message.slice(0, 2045) + '...';
+        }
+        
+        return fetch(location.origin + `/api/game/${room}/chat`, {
+            method: "PUT",
+            body: JSON.stringify({
+                text: message,
+                color, 
+                icon, 
+                user, 
+                character, 
+                title: character || !characterName ? title : `${title} (${characterName})`,
+                hidden: whisper,
+                recipients
+            }),
+            headers: {
+                'x-authorization': `Bearer ${token}`, 'content-type': 'application/json'
+            }
+        });
+    } catch (e) {
+        console.warn(e);
+        message = `**${title}**\n\n` + message;
+
+        if (message.length > 2048) {
+            message = message.slice(0, 2045) + '...';
+        }
+        
+        await speakAs(characterName);
+        sendChatText(message);
+    }
+}
+
+async function updateHpBar({characterName, hp, maxHp, tempHp}) {
+    const room = getRoom();
+    const token = await getAccessToken();
+    const character = await getCharacter(characterName);
+    if (!character) {
+        console.error(`No character found with name ${characterName}`)
+        return
+    }
+    return fetch(location.origin + `/api/game/${room}/character/${character}`, {
+        method: "PATCH",
+        body: JSON.stringify({
+            character: {
+                updateAt: Date.now(),
+                attributeBarMax: maxHp + tempHp,
+                attributeBarCurrent: hp + tempHp,
+
+            }
+        }),
+        headers: {
+            'x-authorization': `Bearer ${token}`, 'content-type': 'application/json'
+        }
+    });
```

Comment:
Not sure if it would be required to check in case of error, clear the character cache and try again (i.e: the character was deleted). But I'm assuming here that characters are not tokens, and tokens are always links to characters on Astral? (sorry don't know how it works).
In which case, I'd say this is fine and clearing character cache would be overengineering it.
Up to you if you think it's a use case that would happen often (i.e: a character with a matching name getting deleted and recreated with a different id).

Reply:
From what I observed tokens are always linked to characters on Astral. Either way the `getCharacters` method tries to look into the cache and if not found tries to refresh the cache from server. An edge case I see here may be if a character is deleted, but still available in cache, then trying to send a message as that character or updating it may error.